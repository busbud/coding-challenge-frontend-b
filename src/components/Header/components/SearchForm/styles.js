import styled from "styled-components";

import { BREAKPOINTS } from "utils/constants";

export const Form = styled.div`
  background-color: var(--white);
  border-radius: 4px;
  margin-top: 16px;
  padding: 16px;

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    align-items: center;
    display: flex;
    padding: 12px 16px;
  }
`;

export const FormControl = styled.div`
  &:not(:last-of-type) {
    margin-bottom: 16px;
  }

  @media screen and (min-width: ${BREAKPOINTS.tablet}) {
    flex: 1;
    margin-bottom: 0;

    &:not(:last-of-type) {
      margin-right: 16px;
      margin-bottom: 0;
    }
  }
`;
