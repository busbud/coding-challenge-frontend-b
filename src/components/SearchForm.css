.SearchForm .datepicker {
  padding: 0 1em; }
