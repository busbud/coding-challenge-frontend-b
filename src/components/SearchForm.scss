.SearchForm {
  .datepicker {
    padding: 0 1.75em;
  }
}