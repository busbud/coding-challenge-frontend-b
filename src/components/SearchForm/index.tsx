import * as React from 'react'
import { Trans } from 'react-i18next'

interface SearchFormProps {
  buttonClick: (ev: React.MouseEvent<HTMLButtonElement>) => void
}

interface SearchFormState {
  from: string,
  to: string,
  date: string,
  passengers: string
}

export default class SearchForm extends React.Component<SearchFormProps, SearchFormState> {
  constructor (props: SearchFormProps) {
    super(props)
    this.state = {
      from: 'New York',
      to: 'Montreal',
      date: '2019-08-02',
      passengers: '1 adult'
    }
  }

  render () {
    return (
      <section className='search'>
        <div className='search-container container'>
          <div className='row justify-content-center align-items-end'>
            <div className='col-lg-2'>
              <label htmlFor='from'>
                <Trans>Departure</Trans>
              </label>
              <input type='text' value={this.state.from} name='from' className='form-input search-from' aria-label='from' readOnly/>
            </div>
            <div className='col-lg-2'>
              <label htmlFor='to'><Trans>Arrival</Trans></label>
              <input type='text' value={this.state.to} name='to' className='form-input search-to' aria-label='to' readOnly/>
            </div>
            <div className='col-lg-3'>
              <label htmlFor='from'>Date</label>
              <input type='date' value={this.state.date} name='date' className='form-input' arial-label='date' readOnly/>
            </div>
            <div className='col-lg-2'>
              <label htmlFor='passenger'><Trans>Passengers</Trans></label>
              <input type='text' value={this.state.passengers} name='passengers' className='form-input' arial-label='passenger' readOnly/>
            </div>
            <div className='col-lg-2'>
              <button type='button' className='form-button' onClick={this.props.buttonClick}><Trans>search</Trans></button>
            </div>
          </div>
        </div>
      </section>
    )
  }
}
