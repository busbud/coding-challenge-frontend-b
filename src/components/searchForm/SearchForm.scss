.SearchForm{
    position: fixed;
    left: 0;
    right: 0;
    top: 30px;
    display: flex;
    align-items: center;
    justify-content: center;
    padding-top: 30px;
    padding-bottom: 30px;
    background-color: #287FC6;
    box-shadow: 0 8px 6px -4px #9E9E9E;



    .text-field {
        margin-right: 20px;
        label + div{
            color: white;
        }
        
        label {
            color: white;
        } 
        fieldset {
            border-color: white;
        }
    }

    button {
        color: white;
        background-color: #ffc107;
        &:hover {
            background-color: #ff9800
        }
    }
}

@media screen and (max-width:500px) {
    .SearchForm {
        position: static;
        flex-direction: column;
        .text-field {
            margin-right: 0;
            min-width: 300px;
        }
        button {
            width: 300px;
            margin-top: 20px;
        }
    }
}

@media screen and (min-width:500px) and (max-width:800px) {
    .SearchForm {

        flex-wrap: wrap;
        .text-field {
            margin-right: 0;
            min-width: 350px;
        }
        button {
            width: 350px;
            margin-top: 20px;
        }
    }
}
