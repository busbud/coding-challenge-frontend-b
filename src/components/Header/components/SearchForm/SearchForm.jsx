import React, { useState } from "react";
import PropTypes from "prop-types";
import { useTranslation } from "react-i18next";

import Select from "commons/Select";
import DatePicker from "commons/DatePicker";
import Button from "commons/Button";

import { Form, FormControl } from "./styles";

const SearchForm = ({ onSearch, ...otherProps }) => {
  const { t } = useTranslation();
  const [originCity, setOriginCity] = useState("");
  const [destinationCity, setDestinationCity] = useState("");
  const [passengers, setPassengers] = React.useState("");
  const [outboundDate, setOutboundDate] = React.useState(null);

  const handleOriginCityChange = (value) => {
    setOriginCity(value);
  };

  const handleDestinationCityChange = (value) => {
    setDestinationCity(value);
  };

  const handlePassengersChange = (value) => {
    setPassengers(value);
  };

  const handleDateChange = (date, dateString) => {
    setOutboundDate(dateString);
  };

  const handleSearch = () => {
    onSearch({
      originCity,
      destinationCity,
      outboundDate,
      passengers,
    });
  };

  return (
    <Form {...otherProps}>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={originCity}
          onChange={handleOriginCityChange}
          data-testid="origin-city-select"
        >
          <Select.Option value="">
            <em>{t("originCity")}</em>
          </Select.Option>
          <Select.Option value="f2m673">Québec</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={destinationCity}
          onChange={handleDestinationCityChange}
          data-testid="destination-city-select"
        >
          <Select.Option value="">
            <em>{t("destinationCity")}</em>
          </Select.Option>
          <Select.Option value="f25dvk">Montréal</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <Select
          style={{ width: "100%" }}
          defaultValue={passengers}
          onChange={handlePassengersChange}
          data-testid="passengers-select"
        >
          <Select.Option value="">
            <em>{t("passengers")}</em>
          </Select.Option>
          <Select.Option value="1">1 {t("adult")}</Select.Option>
        </Select>
      </FormControl>
      <FormControl>
        <DatePicker
          style={{ width: "100%" }}
          onChange={handleDateChange}
          data-testid="outbound-date-picker"
          placeholder="Outbound Date"
        />
      </FormControl>
      <FormControl>
        <Button
          variant="contained"
          color="primary"
          onClick={handleSearch}
          data-testid="search-button"
        >
          {t("search")}
        </Button>
      </FormControl>
    </Form>
  );
};

SearchForm.propTypes = {
  /**
   * The search event trigger.
   */
  onSearch: PropTypes.func.isRequired,
};

export default SearchForm;
