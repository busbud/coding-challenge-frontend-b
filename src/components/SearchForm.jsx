import React from "react";
import 'flatpickr/dist/themes/material_green.css'
import Flatpickr from 'react-flatpickr'
import './SearchForm.css'
import utils from "../utils/utils";

class SearchForm extends React.Component {

  date = this.props.search.date;
  origin = utils.geohashToName(this.props.search.origin);
  destination = utils.geohashToName(this.props.search.destination);

  cityInput(city, onChange) {
    return (
      <div className="field-body is-small">
        <div className="field">
          <p className="control">
            <input
              type="text"
              disabled
              className="input is-small"
              onChange={onChange}
              value={city}
            />
          </p>
        </div>
      </div>
    );
  }

  datePicker = (
    <div className="control datepicker">
      <Flatpickr
        className="input is-small" value={this.date}
        onChange={(newDate) => {
          this.date = newDate[0]
        }}
      />
    </div>
  );

  submitButton = (
    <div className="control">
      <button
        onClick={() => {
          this.props.onSearchClick(this.props.search.origin, this.props.search.destination, this.date)
        }}
        className="button is-primary is-small">Search</button>
    </div>
  );

  render() {
    return (
      <div className="SearchForm field is-horizontal">
        <div className="field-label is-small">
          <label className="label">From</label>
        </div>
        {this.cityInput(this.origin, this.props.onOriginChange)}
        <div className="field-label is-small">
          <label className="label">To</label>
        </div>
        {this.cityInput(this.destination, this.props.onDestinationChange)}

        {this.datePicker}

        {this.submitButton}
      </div>
    );
  }
};

export default SearchForm;
