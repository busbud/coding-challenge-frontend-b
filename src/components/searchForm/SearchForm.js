import React, { Component } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { connect } from 'react-redux';
import { wait } from '../../helpers/wait';
import { cleanStore, searchRequestSuccess } from '../../actions/search.actions';
import { URL_INIT_SEARCH, URL_POLL_SEARCH, HEARDERS } from '../../constants/applications';
import './SearchForm.scss';

export class SearchForm extends Component {
    constructor() {
        super();
        this.state = {
            departure: "New York",
            arrival: "Montreal",
            date: "2019-08-02",
            passenger: "1 Adult",
            departures: []
        }
        this.handleSearch = this.handleSearch.bind(this);
    }

    async handleSearch() {
        this.props.cleanStore();
        try {
            const response = await fetch(URL_INIT_SEARCH, HEARDERS)
            if(response.status === 200) {
                await this.fetchWithIndex(URL_POLL_SEARCH, 10);
            }
        } catch (err) {
            console.log('Init search failed');
        }
    }
    
    async fetchWithIndex(url, index){
        try {
            // const WITH_INDEX = `${url}${index}`;
            index += 10;
            const response = await fetch(url, HEARDERS);
            const data = await response.json();
            this.props.searchRequestSuccess(data)
    
            if(!data.complete){
                await wait(3000);
                fetchWithIndex(URL_POLL_SEARCH, index);
            }
    
        } catch (err) {
            console.log('Poll search failed');
        }
    }

    render() {
        return (
            <div className="SearchForm">
                <TextField
                    label="Leaving from"
                    className="text-field"
                    value={this.state.departure}
                    margin="normal"
                    variant="outlined"
                />
                <TextField
                    label="Going to"
                    className="text-field"
                    value={this.state.arrival}
                    margin="normal"
                    variant="outlined"
                />
                <TextField
                    label="Date"
                    className="text-field"
                    value={this.state.date}
                    margin="normal"
                    variant="outlined"
                />
                <TextField
                    label="passenger"
                    className="text-field"
                    value={this.state.passenger}
                    margin="normal"
                    variant="outlined"
                />
                <Button
                    variant="contained"
                    color="primary"
                    onClick={this.handleSearch}
                >
                    Search
                </Button>
            </div>
        )
    }
}

const mapDispatchToProps = {
    cleanStore,
    searchRequestSuccess
}

export default connect(undefined, mapDispatchToProps)(SearchForm);