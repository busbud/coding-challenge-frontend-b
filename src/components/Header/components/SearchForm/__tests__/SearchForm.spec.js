import React from "react";
import flushPromises from "flush-promises";
import { render, fireEvent, act } from "@testing-library/react";

import "i18n";

import SearchForm from "../";

// Date.now = jest.fn(() => 1610206111944); //01.09.2021

const mockedOnSearch = jest.fn();

describe("<SearchForm />", () => {
  it("should render properly", () => {
    const { container } = render(<SearchForm onSearch={mockedOnSearch} />);
    expect(container).toMatchSnapshot();
  });

  it("should handle the onSearch", async () => {
    const { baseElement, queryByTestId, queryByText } = render(
      <SearchForm onSearch={mockedOnSearch} />
    );
    const originCitySelect = queryByTestId("origin-city-select");
    const destinationCitySelect = queryByTestId("destination-city-select");
    const passengersSelect = queryByTestId("passengers-select");
    const datePicker = queryByTestId("outbound-date-picker");

    await act(async () => {
      fireEvent.mouseDown(originCitySelect.firstElementChild);
      await flushPromises();
      fireEvent.click(queryByText("Québec"));
    });

    await act(async () => {
      fireEvent.mouseDown(destinationCitySelect.firstElementChild);
      await flushPromises();
      fireEvent.click(queryByText("Montréal"));
    });

    await act(async () => {
      fireEvent.mouseDown(passengersSelect.firstElementChild);
      await flushPromises();
      fireEvent.click(queryByText("1 adult"));
    });

    await act(async () => {
      fireEvent.mouseDown(datePicker);
      await flushPromises();
      fireEvent.change(datePicker, { target: { value: "2021-01-10" } });
      fireEvent.click(baseElement.querySelector(".ant-picker-cell-selected"));
    });

    fireEvent.click(queryByTestId("search-button"));

    expect(mockedOnSearch).toHaveBeenCalledWith({
      originCity: "f2m673",
      destinationCity: "f25dvk",
      outboundDate: "2021-01-10",
      passengers: "1",
    });
  });
});
