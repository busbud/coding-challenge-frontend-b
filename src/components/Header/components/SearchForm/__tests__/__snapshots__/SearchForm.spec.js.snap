// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm /> should render properly 1`] = `
<div>
  <div
    class="sc-hKgILt yjWhv"
  >
    <div
      class="sc-eCssSg dFltHD"
    >
      <div
        class="ant-select sc-bdfBwQ bLtaVd ant-select-single ant-select-show-arrow"
        data-testid="origin-city-select"
        style="width: 100%;"
      >
        <div
          class="ant-select-selector"
        >
          <span
            class="ant-select-selection-search"
          >
            <input
              aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
              aria-autocomplete="list"
              aria-controls="rc_select_TEST_OR_SSR_list"
              aria-haspopup="listbox"
              aria-owns="rc_select_TEST_OR_SSR_list"
              autocomplete="off"
              class="ant-select-selection-search-input"
              id="rc_select_TEST_OR_SSR"
              readonly=""
              role="combobox"
              style="opacity: 0;"
              type="search"
              unselectable="on"
              value=""
            />
          </span>
          <span
            class="ant-select-selection-item"
          >
            <em>
              Origin City
            </em>
          </span>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="sc-eCssSg dFltHD"
    >
      <div
        class="ant-select sc-bdfBwQ bLtaVd ant-select-single ant-select-show-arrow"
        data-testid="destination-city-select"
        style="width: 100%;"
      >
        <div
          class="ant-select-selector"
        >
          <span
            class="ant-select-selection-search"
          >
            <input
              aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
              aria-autocomplete="list"
              aria-controls="rc_select_TEST_OR_SSR_list"
              aria-haspopup="listbox"
              aria-owns="rc_select_TEST_OR_SSR_list"
              autocomplete="off"
              class="ant-select-selection-search-input"
              id="rc_select_TEST_OR_SSR"
              readonly=""
              role="combobox"
              style="opacity: 0;"
              type="search"
              unselectable="on"
              value=""
            />
          </span>
          <span
            class="ant-select-selection-item"
          >
            <em>
              Destination City
            </em>
          </span>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="sc-eCssSg dFltHD"
    >
      <div
        class="ant-select sc-bdfBwQ bLtaVd ant-select-single ant-select-show-arrow"
        data-testid="passengers-select"
        style="width: 100%;"
      >
        <div
          class="ant-select-selector"
        >
          <span
            class="ant-select-selection-search"
          >
            <input
              aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
              aria-autocomplete="list"
              aria-controls="rc_select_TEST_OR_SSR_list"
              aria-haspopup="listbox"
              aria-owns="rc_select_TEST_OR_SSR_list"
              autocomplete="off"
              class="ant-select-selection-search-input"
              id="rc_select_TEST_OR_SSR"
              readonly=""
              role="combobox"
              style="opacity: 0;"
              type="search"
              unselectable="on"
              value=""
            />
          </span>
          <span
            class="ant-select-selection-item"
          >
            <em>
              Passengers
            </em>
          </span>
        </div>
        <span
          aria-hidden="true"
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-select-suffix"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="sc-eCssSg dFltHD"
    >
      <div
        class="ant-picker sc-gsTCUz jhttyt"
        style="width: 100%;"
      >
        <div
          class="ant-picker-input"
        >
          <input
            autocomplete="off"
            data-testid="outbound-date-picker"
            placeholder="Outbound Date"
            readonly=""
            size="12"
            title=""
            value=""
          />
          <span
            class="ant-picker-suffix"
          >
            <span
              aria-label="calendar"
              class="anticon anticon-calendar"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="calendar"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="sc-eCssSg dFltHD"
    >
      <button
        class="sc-dlfnbm ikhxZP"
        color="primary"
        data-testid="search-button"
        type="button"
      >
        Search
      </button>
    </div>
  </div>
</div>
`;
